tensorflow/core/lib/random/random_distributions.h

```diff
@@ -85,56 +129,84 @@ class UniformDistribution<Generator, Eigen::half> {
   }
 };
 
-template <class Generator>
-class UniformDistribution<Generator, bfloat16> {
+template <class Generator, bool IsVec>
+class UniformDistribution<Generator, bfloat16, IsVec> {
  public:
   // The number of elements that will be returned.
-  static constexpr int kResultElementCount = Generator::kResultElementCount;
+  // Set the number to be Eigen packet size of type at least, so computations
+  // can be vectorized using SIMD on CPU.
+  static constexpr int kVectorLength = std::max(
+      static_cast<const int>(Eigen::internal::packet_traits<bfloat16>::size),
+      Generator::kResultElementCount);
+  static constexpr int kResultElementCount =
+      IsVec ? kVectorLength : Generator::kResultElementCount;
   // Cost of generation of a single element (in cycles).
   static constexpr int kElementCost = 3;
   // Indicate that this distribution may take variable number of samples
   // during the runtime.
   static constexpr bool kVariableSamplesPerOutput = false;
   typedef Array<bfloat16, kResultElementCount> ResultType;
   typedef bfloat16 ResultElementType;
+  // Helper definition for the format function.
+  typedef bfloat16 (*FormatFunc)(uint16);
 
   PHILOX_DEVICE_INLINE
   ResultType operator()(Generator* gen) {
+#ifdef __CUDA_ARCH__
+    static_assert(!IsVec, "Can't vectorize Distribution on GPU");
```

Comment:
Then shouldn't there be a non-vectorized specialization for GPU so we never run into this?

Reply:
Yes, this `IsVec` flag is specified in kernel registry and only CPU has registered it. GPU should never run with this flag so far.

Reply:
I think it should still fall back to the scalar version - or otherwise implement the vectorized version in a similar way.